File: practice_problems/fcp/frequency_counter_num.go
package main

import (
	"fmt"
)

func main() {
	// fmt.Println(sameFCP([]int{1, 2, 3, 2, 5}, []int{9, 1, 4, 4, 25}))
	result := anagrams("anagram", "nagaram")
	fmt.Println(result)
}

func anagrams(s1, s2 string) bool {
	if len(s1) != len(s2) {
		return false
	}
	fmt.Println(s1)
	// s1 = strings.ToLower(s1)
	// s2 = strings.ToLower(s2)

	fc1 := make(map[rune]int)
	fc2 := make(map[rune]int)

	for _, char := range s1 {
		fc1[char] += 1
	}
	for _, char := range s2 {
		fc2[char] += 1
	}

	for key := range fc1 {
		if _, ok := fc1[key]; !ok {
			return false
		}
		if fc1[key] != fc2[key] {
			return false
		}
	}
	return true
}

func sameFCP(a1, a2 []int) bool {
	if len(a1) != len(a2) {
		return false
	}

	fc1 := make(map[int]int)
	fc2 := make(map[int]int)

	for _, val := range a1 {
		fc1[val] += 1
	}
	for _, val := range a2 {
		fc2[val] += 1
	}

	for key := range fc1 {
		if _, ok := fc2[key*key]; !ok {
			return false
		}
		if fc1[key] != fc2[key*key] {
			return false
		}
	}
	return true
}

func same(a1, a2 []int) bool {
	if len(a1) != len(a2) {
		return false
	}
	// loop a1
	// check square value of a1 is present in a2 looping a2
	// if yes return index of a2 value
	// splice the index value from a2 and continue loop
	idx := -1
	for i := range a1 {
		for j := range a2 {
			if a2[j] == a1[i]*a1[i] {
				idx = j
				break
			}
		}
		if idx == -1 {
			return false
		}
		a2 = append(a2[:idx], a2[idx+1:]...)
	}
	return true
}

File: practice_problems/fcp/frequency_counter_num.js
console.log("Frequency Counter Pattern")

// a) Understand the problem
// - Restate the problem
// -> All values in array one must have corresponding squared values in array two and should
// match the frequency and maintaining order is not necessary
// - Inputs that go into this problem
// -> An array of integers
// - Outputs from the solution
// -> An array of squared integers
// - Is the information enough
// -> yes
// - Labeling the data
// - Create a function that takes two input arrays and returns boolean

// b) Explore concrete example
// - Start with simple examples
// -> arr1=[1, 2, 3, 4] Arr2=[1, 4, 9, 16] o/p=true

// c) Break it down
// - Write steps to solve the problem
// -> create a function that takes two integer arrays as input and returns boolean
// -> check if the length of input arrays are same if yes continue or else return false
// -> loop the first array and check each element in the first array has the corresponding squared element
// -> if yes then remove that element from 1st array and repeat above process until first array is empty
// -> if no then return false
// -> return true after the loop

// d) Slove the problem
function same(arr1, arr2) {
  if (arr1.length != arr2.length) {
    return false
  }
  for (let i = 0; i < arr1.length; i++) {
    let idx = arr2.indexOf(arr1[i] ** 2)
    if (idx == -1) {
      return false
    }
    arr2.splice(idx, 1)
  }
  return true
}

// console.log(same([1, 2, 3, 4], [16, 9, 4, 1]))

// Following frequency counter pattern
function sameFCP(a1, a2) {
  if (a1.length != a2.length) {
    return false
  }

  let fc1 = {}
  let fc2 = {}

  console.log(a1)
  console.log(a2)

  for (let val of a1) {
    fc1[val] = (fc1[val] || 0) + 1
  }
  for (let val of a2) {
    fc2[val] = (fc2[val] || 0) + 1
  }

  console.log(fc1)
  console.log(fc2)

  for (let key in fc1) {
    if (!(key ** 2 in fc2)) {
      return false
    }
    if (fc2[key ** 2] != fc1[key]) {
      return false
    }
  }
  return true
}

console.log(sameFCP([1, 2, 3, 4], [16, 9, 4, 1]))